File: Laborator/lab05 - Backtracking/skel/java/task-1/src/Main.java
import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;

public class Main {
    static class Task {
        public final static String INPUT_FILE = "in";
        public final static String OUTPUT_FILE = "out";

        int n, k;

        public void solve() {
            readInput();
            writeOutput(getResult());
        }

        private void readInput() {
            try {
                Scanner sc = new Scanner(new File(INPUT_FILE));
                n = sc.nextInt();
                k = sc.nextInt();
                sc.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        private void writeOutput(ArrayList<ArrayList<Integer>> result) {
            try {
                PrintWriter pw = new PrintWriter(new File(OUTPUT_FILE));
                pw.printf("%d\n", result.size());
                for (ArrayList<Integer> arr : result) {
                    for (int i = 0; i < arr.size(); i++) {
                        pw.printf("%d%c", arr.get(i), i + 1 == arr.size() ?
                                '\n' : ' ');
                    }
                }
                pw.close();
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        }

        private void bkt(int step, int stop, ArrayList<Integer> domain, ArrayList<Integer> solution,
                         Set<Integer> visited, ArrayList<ArrayList<Integer>> all) {
            if (step == stop) {
                all.add(new ArrayList<>(solution));
                return;
            }

            for (int i = 0; i < n; i++) {
                if (!visited.contains(domain.get(i))) {
                    visited.add(domain.get(i));
                    solution.add(step, domain.get(i));
                    bkt(step + 1, stop, domain, solution, visited, all);
                    solution.remove(solution.size() - 1);
                    visited.remove(domain.get(i));
                }
            }
        }

        private ArrayList<ArrayList<Integer>> getResult() {
            ArrayList<ArrayList<Integer>> all = new ArrayList<>();

            // TODO: Construiti toate aranjamentele de N luate cate K ale
            // multimii {1, ..., N}.
            //
            // Pentru a adauga un nou aranjament:
            //   ArrayList<Integer> aranjament;
            //   all.add(aranjament);
            //

            Set<Integer> visited = new HashSet<Integer>(0);
            ArrayList<Integer> solution = new ArrayList<>(k);
            ArrayList<Integer> domain = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                domain.add(i + 1);
            }

            bkt(0, k, domain, solution, visited, all);

            return all;
        }
    }

    public static void main(String[] args) {
        new Task().solve();
    }
}

File: Laborator/lab05 - Backtracking/skel/cpp/task-1/main.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <unordered_set>
using namespace std;

class Task {
public:
    void solve() {
        read_input();
        print_output(get_result());
    }

private:
    int n, k;

    void read_input() {
        ifstream fin("in");
        fin >> n >> k;
        fin.close();
    }

    void bkt(int step, int stop, vector<int> &domain, vector<int> &solution, unordered_set<int> &visited, vector<vector<int> > &all) {
        if (step == stop) {
            all.push_back(solution);
            return;
        }

        for (int i = 0; i < domain.size(); i++) {
            if (visited.find(domain[i]) == visited.end()) {
                visited.insert(domain[i]);
                solution[step] = domain[i];
                bkt(step + 1, stop, domain, solution, visited, all);
                visited.erase(domain[i]);
            }
        }
    }

    vector<vector<int> > get_result() {
        vector<vector<int> > all;

        // TODO: Construiti toate aranjamentele de N luate cate K ale
        // multimii {1, ..., N}.
        //
        // Pentru a adauga un nou aranjament:
        //     vector<int> aranjament;
        //     all.push_back(aranjament);

        unordered_set<int> visited;
        vector<int> solution(k);
        vector<int> domain(n);
        for (int i = 0; i < n; i++) {
            domain[i] = i + 1;
        }

        bkt(0, k, domain, solution, visited, all);
        return all;
    }

    void print_output(const vector<vector<int> >& result) {
        ofstream fout("out");
        fout << result.size() << '\n';
        for (size_t i = 0; i < result.size(); i++) {
            for (size_t j = 0; j < result[i].size(); j++) {
                fout << result[i][j] << (j + 1 != result[i].size() ? ' ' : '\n');
            }
        }
        fout.close();
    }
};

// [ATENTIE] NU modifica functia main!
int main() {
    // * se aloca un obiect Task pe heap
    // (se presupune ca e prea mare pentru a fi alocat pe stiva)
    // * se apeleaza metoda solve()
    // (citire, rezolvare, printare)
    // * se distruge obiectul si se elibereaza memoria
    auto* task = new (std::nothrow) Task{}; // hint: cppreference/nothrow
    if (!task) {
        std::cerr << "new failed: WTF are you doing? Throw your PC!\n";
        return -1;
    }
    task->solve();
    delete task;
    return 0;
}
